logs_ingest/logging.py

```diff
@@ -15,28 +15,54 @@
 import logging
 import os
 
+LOG_THROTTLING_LIMIT_PER_CALLER = 10
+log_call_count = dict()
+
 script_directory = os.path.dirname(os.path.realpath(__file__))
 version_file_path = os.path.join(script_directory, "version.txt")
 with open(version_file_path) as version_file:
     _version = version_file.readline()
     _version_tag = f"[{_version}] "
 
 
-def exception(msg, *args, **kwargs):
+def exception(msg, caller: str, *args, **kwargs):
+    if check_if_caller_exceeded_limit(caller):
+        return
     logging.exception(_version_tag + msg, *args, **kwargs)
 
 
-def error(msg, *args, **kwargs):
+def error(msg, caller: str, *args, **kwargs):
+    if check_if_caller_exceeded_limit(caller):
+        return
     logging.error(_version_tag + msg, *args, **kwargs)
 
 
-def warning(msg, *args, **kwargs):
+def warning(msg, caller: str, *args, **kwargs):
+    if check_if_caller_exceeded_limit(caller):
+        return
     logging.warning(_version_tag + msg, *args, **kwargs)
 
 
 def info(msg, *args, **kwargs):
     logging.info(_version_tag + msg, *args, **kwargs)
 
 
-def debug(msg, *args, **kwargs):
-    logging.debug(_version_tag + msg, *args, **kwargs)
\ No newline at end of file
+def debug(msg, caller: str, *args, **kwargs):
+    if check_if_caller_exceeded_limit(caller):
+        return
+    logging.debug(_version_tag + msg, *args, **kwargs)
+
+
+def check_if_caller_exceeded_limit(caller):
+    log_calls_left = LOG_THROTTLING_LIMIT_PER_CALLER - log_call_count.get(caller, 0)
+
+    if log_calls_left == 0:
+        log_call_count[caller] = log_call_count.get(caller, 0) + 1
+        logging.warning(_version_tag + f"Logging calls from caller '{caller}' exceeded the throttling limit of"
+                        f" {LOG_THROTTLING_LIMIT_PER_CALLER}. Further logs from this caller will be discarded")
+
+    caller_exceeded_limit = log_calls_left <= 0
+    if not caller_exceeded_limit:
+        log_call_count[caller] = log_call_count.get(caller, 0) + 1
```

Comment:
The same value is extracted from the dict three times. Maybe we should just extract it once at the beginning of the function and assign it to a variable, e.g. `log_calls_performed`?

Reply:
It's extracted two times at most during one execution but it's probably still better to just assign it, will do.

Reply:
Yeah, I don't mean runtime overhead here but rather code duplication